pkg/controller/controller_test.go

```diff
@@ -0,0 +1,188 @@
+		NativeVolumeClaim: mock.CreatePVC(1, 2, "pvc", "pvcuid", mock.DefaultNS, "pv", v1.ClaimBound),
+	}
+
+	testCase := &testCase{
+		name:               "abnormal_volume_case1",
+		enableNodeWatcher:  false,
+		supportListVolumes: true,
+		fakeNativeObjects: &fakeNativeObjects{
+			MockVolume: abnormalVolume,
+		},
+		wantEvent: true,
+	}
+
+	runTest(t, testCase)
+}
+
+func Test_AbnormalVolumeWithNodeWatcher(t *testing.T) {
+	abnormalVolume := &mock.MockVolume{
+		CSIVolume: &mock.CSIVolume{
+			Volume: &csi.Volume{
+				VolumeId: "abnormalVolume1",
+			},
+			Condition: &csi.VolumeCondition{
+				Abnormal: true,
+				Message:  "Volume not found",
+			},
+		},
+		NativeVolume:      mock.CreatePV(2, "pvc", "pv", mock.DefaultNS, "abnormalVolume1", "pvcuid", &mock.FSVolumeMode, v1.VolumeBound),
+		NativeVolumeClaim: mock.CreatePVC(1, 2, "pvc", "pvcuid", mock.DefaultNS, "pv", v1.ClaimBound),
+	}
+
+	abnormalNodes := &mock.MockNode{
+		NativeNode: mock.CreateNode("node1", ""),
+	}
+
+	testCase := &testCase{
+		name:               "abnormal_volume_case1",
+		enableNodeWatcher:  true,
+		supportListVolumes: true,
+		fakeNativeObjects: &fakeNativeObjects{
+			MockVolume: abnormalVolume,
+			MockNode:   abnormalNodes,
+		},
+		wantEvent: true,
+	}
+
+	runTest(t, testCase)
+}
+
+func Test_NormalVolumeWithoutNodeWatcher(t *testing.T) {
+	normalVolume := &mock.MockVolume{
+		CSIVolume: &mock.CSIVolume{
+			Volume: &csi.Volume{
+				VolumeId: "normalVolume1",
+			},
+			Condition: &csi.VolumeCondition{
+				Abnormal: false,
+				Message:  "",
+			},
+		},
+		NativeVolume:      mock.CreatePV(2, "pvc", "pv", mock.DefaultNS, "normalVolume1", "pvcuid", &mock.FSVolumeMode, v1.VolumeBound),
+		NativeVolumeClaim: mock.CreatePVC(1, 2, "pvc", "pvcuid", mock.DefaultNS, "pv", v1.ClaimBound),
+	}
+
+	testCase := &testCase{
+		name:               "normal_volume_case1",
+		enableNodeWatcher:  false,
+		supportListVolumes: true,
+		fakeNativeObjects: &fakeNativeObjects{
+			MockVolume: normalVolume,
+		},
+		wantEvent: false,
+	}
+
+	runTest(t, testCase)
+}
+
+func Test_AbnormalVolumeWithoutNodeWatcherAndListVolume(t *testing.T) {
+	abnormalVolume := &mock.MockVolume{
+		CSIVolume: &mock.CSIVolume{
+			Volume: &csi.Volume{
+				VolumeId: "abnormalVolume1",
+			},
+			Condition: &csi.VolumeCondition{
+				Abnormal: true,
+				Message:  "Volume not found",
+			},
+		},
+		NativeVolume:      mock.CreatePV(2, "pvc", "pv", mock.DefaultNS, "abnormalVolume1", "pvcuid", &mock.FSVolumeMode, v1.VolumeBound),
+		NativeVolumeClaim: mock.CreatePVC(1, 2, "pvc", "pvcuid", mock.DefaultNS, "pv", v1.ClaimBound),
+	}
+
+	testCase := &testCase{
+		name:               "abnormal_volume_case1",
+		enableNodeWatcher:  false,
+		supportListVolumes: false,
+		fakeNativeObjects: &fakeNativeObjects{
+			MockVolume: abnormalVolume,
+		},
+		wantEvent: true,
+	}
+
+	runTest(t, testCase)
+}
+
+func Test_AbnormalVolumeWithNodeWatcherNoListVolume(t *testing.T) {
```

Comment:
I have a general comments.  I think most of the test cases can be in the same Test_func.  Inside the func, you can have different test cases.  If you look at unit tests in other sidecars, it is usually structured that way.  But this can be done at a later time.

Reply:
Sure